src/WebpackTools/PWADevServer.js

```diff
@@ -0,0 +1,167 @@
+    }),
+    hostnamesById: new GlobalConfig({
+        prefix: 'devhostname-byid',
+        key: x => x
+    }),
+    portsByHostname: new GlobalConfig({
+        prefix: 'devport-byhostname',
+        key: x => x
+    }),
+    async setLoopback(hostname) {
+        debug(`checking if ${hostname} is loopback`);
+        let ip;
+        try {
+            ip = await lookup(hostname);
+        } catch (e) {
+            if (e.code !== 'ENOTFOUND') {
+                throw Error(
+                    debug.errorMsg(
+                        `Error trying to check that ${hostname} is loopback: ${
+                            e.message
+                        }`
+                    )
+                );
+            }
+        }
+        if (ip && (ip.address === '127.0.0.1' || ip.address === '::1')) {
+            debug(`${hostname} already resolves to ${ip.address}!`);
+        } else {
+            debug(
+                `setting ${hostname} loopback in /etc/hosts, may require password...`
+            );
+            return runAsRoot(
+                /* istanbul ignore next: never runs in process */
+                d => require('hostile').set('127.0.0.1', d),
+                hostname
+            );
+        }
+    },
+    async findFreePort() {
+        const inUse = await PWADevServer.portsByHostname.values(Number);
+        debug(`findFreePort(): these ports already in use`, inUse);
+        return findPort({
+            startingPort: 8000,
+            endingPort: 9999,
+            avoid: inUse
+        }).catch(e => {
+            throw Error(
+                debug.errorMsg(
+                    `Unable to find an open port. You may want to delete your database file at ${GlobalConfig.getDbFilePath()} to clear out old developer hostname entries. (Soon we will make this easier and more automatic.) Original error: ${e.toString()}`
+                )
+            );
+        });
+    },
+    async findFreeHostname(identifier, times = 0) {
+        const maybeHostname =
+            identifier + (times ? times : '') + '.local.pwadev';
+        // if it has a port, it exists
+        const exists = await PWADevServer.portsByHostname.get(maybeHostname);
+        if (!exists) {
+            debug(
+                `findFreeHostname: ${maybeHostname} unbound to port and available`
+            );
+            return maybeHostname;
+        } else {
+            debug(`findFreeHostname: ${maybeHostname} bound to port`, exists);
+            return PWADevServer.findFreeHostname(identifier, times + 1);
+        }
+    },
+    async provideDevHost(id) {
+        debug(`provideDevHost('${id}')`);
+        let hostname = await PWADevServer.hostnamesById.get(id);
+        let port;
+        if (!hostname) {
+            [hostname, port] = await Promise.all([
+                PWADevServer.findFreeHostname(id),
+                PWADevServer.findFreePort()
+            ]);
+
+            await PWADevServer.hostnamesById.set(id, hostname);
```

Comment:
Can these 2 `await` calls run in parallel?

Reply:
I don't remember, but since they were both touching the flat file I felt like playing it safe.